arangod/RocksDBEngine/RocksDBMetaCollection.cpp

```diff
@@ -637,64 +839,161 @@ void RocksDBMetaCollection::corruptRevisionTree(uint64_t count, uint64_t hash) {
+        while (it != _revisionInsertBuffers.end() && it->first <= seq) {
           it = _revisionInsertBuffers.erase(it);
         }
       }
-        
+      
       {
         auto it = _revisionRemovalBuffers.begin(); 
-        while (it != _revisionRemovalBuffers.end() && it->first <= beginSeq) {
+        while (it != _revisionRemovalBuffers.end() && it->first <= seq) {
           it = _revisionRemovalBuffers.erase(it);
         }
       }
+    };
+    
+    // place a blocker and add a guard to remove it at the end
+    TransactionId trxId = TransactionId(transaction::Context::makeTransactionId());
+  
+    auto blockerGuard = scopeGuard([&] {  // remove blocker afterwards
+      removeRevisionTreeBlocker(trxId);
+    });
+      
+    TRI_IF_FAILURE("rebuildRevisionTree::sleep") {
+      std::this_thread::sleep_for(std::chrono::milliseconds(RandomGenerator::interval(uint32_t(2000))));
+    }
+    
+    // we now have a blocker in place that will prevent updates from being 
+    // applied
+    rocksdb::SequenceNumber blockerSeq = placeRevisionTreeBlocker(trxId);
+    
+    TRI_IF_FAILURE("rebuildRevisionTree::sleep") {
+      std::this_thread::sleep_for(std::chrono::milliseconds(RandomGenerator::interval(uint32_t(2000))));
     }
 
-    return {};
+    {
+      // we may still have some buffered updates, so let's remove them.
+      std::unique_lock<std::mutex> guard(_revisionTreeLock);
+      removeBufferedUpdatesUpTo(blockerSeq - 1);
+
+      _revisionTreeApplied = blockerSeq - 1;
+    }
+  
+    // unlock the collection again
+    lockGuard.fire();
+    
+    TRI_IF_FAILURE("rebuildRevisionTree::sleep") {
+      std::this_thread::sleep_for(std::chrono::milliseconds(RandomGenerator::interval(uint32_t(2000))));
+    }
+
+    // now rebuild the tree from the collection data, using a
+    // snapshot that will be acquired inside `revisionTreeFromCollection`.
+    // note that the snapshot may have a sequence number > blockerSeq!
+    auto result = revisionTreeFromCollection(false);
+    if (result.fail()) {
+      return result.result(); 
+    }
+
+    auto&& [newTree, beginSeq] = result.get();
+
+    TRI_ASSERT(blockerSeq <= beginSeq);
+
+    // update our blocker to the sequence number from the snapshot 
+    _meta.updateBlocker(trxId, beginSeq);
+    
+    TRI_IF_FAILURE("rebuildRevisionTree::sleep") {
+      std::this_thread::sleep_for(std::chrono::milliseconds(RandomGenerator::interval(uint32_t(2000))));
+    }
+
+    // we need to set a timeout for this operation, because if it does not complete within 
+    // reasonable bounds, we need to relinquish the thread that is kept busy waiting here.
+    auto const endTime = std::chrono::steady_clock::now() + std::chrono::minutes(2);
+
+    // wait until all blockers before the snapshot have been removed
+    while (true) {
+      std::unique_lock<std::mutex> guard(_revisionTreeLock);
+      
+      if (_revisionTreeApplied < beginSeq) {
+        _revisionTreeApplied = beginSeq;
+      }
```

Comment:
I think we should not change the old tree's `_revisionTreeApplied`. It is probably intended to prevent new blockers older than beginSeq from being inserted, but since we already track the max sequence number in `placeBlocker` this should not be necessary (but it is probably also not problematic).

Reply:
We decided to keep it as is since it doesn't hurt and we want to finalize this.